common/lanelet2_extension/lib/SignalizedIntersection.cpp

```diff
@@ -0,0 +1,191 @@
+    {
+        return llt.regulatoryElementsAs<CarmaTrafficSignal>();
+    }
+
+    void SignalizedIntersection::addLanelet(Lanelet lanelet, IntersectionSection section)
+    {
+        switch (section)
+        {
+            case IntersectionSection::ENTRY:
+                parameters()[CarmaRoleNameString::IntersectionEntry].emplace_back(RuleParameter(lanelet));
+                break;
+            case IntersectionSection::EXIT:
+                parameters()[CarmaRoleNameString::IntersectionExit].emplace_back(RuleParameter(lanelet));
+                break;
+            case IntersectionSection::INTERIOR:
+                parameters()[CarmaRoleNameString::IntersectionInterior].emplace_back(RuleParameter(lanelet));
+                break;
+            default:
+                throw std::invalid_argument("Invalid section is passed as IntersectionSection");
+        }
+        
+        section_lookup[lanelet.id()] = section;
+    }
+    
+    namespace {
+
+        template <typename T>
+        bool findAndErase(const T& primitive, RuleParameters* member)
+        {
+            if (member == nullptr)
+            {
+                std::cerr << __FUNCTION__ << ": member is null pointer";
+                return false;
+            }
+
+            auto it = std::find(member->begin(), member->end(), RuleParameter(primitive));
+            
+            if (it == member->end())
+            {
+                return false;
+            }
+
+            member->erase(it);
+
+            return true;
+        }
+
+        class GetIdVisitor : public RuleParameterVisitor {
+            public:
+                static Id id(const ConstRuleParameter& param) {
+                    GetIdVisitor visitor;
+                    boost::apply_visitor(visitor, param);
+                    return visitor.id_;
+                }
+                template <typename PrimT>
+                void appendID(const PrimT& p) {
+                    id_ = p.id();
+                }
+
+                void operator()(const ConstPoint3d& p) override { appendID(p); }
+                void operator()(const ConstLineString3d& l) override { appendID(l); }
+                void operator()(const ConstPolygon3d& p) override { appendID(p); }
+                void operator()(const ConstWeakLanelet& ll) override {
+                    if (!ll.expired()) {
+                    appendID(ll.lock());
+                    }
+                }
+                void operator()(const ConstWeakArea& ar) override {
+                    if (!ar.expired()) {
+                    appendID(ar.lock());
+                    }
+                }
+
+            private:
+                Id id_{};
+        };
+        
+    } //namespace
+
+    bool SignalizedIntersection::removeLanelet(const Lanelet& llt)
+    {
+        // if successfully found and erased, then return true, else false
+        auto iter = section_lookup.find(llt.id());
```

Comment:
Another lookup is occurring, but the previous lookups also still occur in findAndErase? So what was accomplished here? Looking for some clarification.

Reply:
Ah I see now. I don't think speed is a huge concern here given the likely number of lanelets for each intersection so this is fine.